### PR TITLE
DrivAerML: Decaying Weight Decay Schedule

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DrivAerML champion uses NO weight decay (WD=0, val=3.833%). Previous experiments with **constant WD=1e-3 got 4.44%** — worse than baseline. The problem with constant WD is well understood: it penalizes large weights equally throughout training, but in late training the model needs precise, possibly large weights to fit high-curvature surface regions.

A **decaying weight decay schedule** provides regularization during early training (helping the model explore and avoid sharp minima) while annealing to zero in late training (allowing unconstrained convergence to the deepest basin). This is conceptually similar to how learning rate decay works — strong regularization early, none late.

**Key comparison:**
- WD=1e-3 constant: **4.44%** — WD too strong, maintained throughout, penalizes late convergence
- WD=0 (champion): **3.833%** — no regularization benefit at all
- **This experiment:** WD starts small and decays to zero — gets early regularization benefit without late-training penalty

---

## Baseline

Current best DrivAerML (PR #3072, `eren/dm-ema-9995-gc05`):
- `val_primary/surface_rel_l2_pct`: **3.833%** at epoch 511
- test: 4.685%
- W&B run: `ncl1dh88`
- Config: AdamW lr=5e-4, gc=0.5, NO WD, 4L/512d/8H, EMA=0.9995, T_max=30, 394 batches/epoch

**External target:** <3.71% (AB-UPT). Gap: ~0.123 pp.

**Target: beat 3.833%.**

---

## Known Bug Fix

There is a `primary_metric_key` shadowing bug in `train.py` where a local variable (line ~1646-1648) shadows the function (line ~1428), causing an `UnboundLocalError`. Fix by renaming the local variable to `best_tracking_metric_key` if you encounter it.

---

## Implementation

1. Add `--wd-decay-epochs` flag (integer, default 0 meaning constant WD / no decay)
2. When set to N > 0, linearly decay the weight decay from its initial value to 0 over N epochs
3. At epoch e: `current_wd = initial_wd * max(0, 1 - e/N)`
4. After epoch N, WD = 0 for the remainder of training
5. Update the optimizer's WD parameter group at each epoch
6. Log `current_wd` to W&B each epoch

---

### Trial 1 — WD=1e-4 decaying to 0 over 300 epochs

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --weight-decay 1e-4 \
  --wd-decay-epochs 300 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb_group dm-decaying-wd
```

### Trial 2 — WD=5e-5 decaying to 0 over 300 epochs

Same as Trial 1 but with `--weight-decay 5e-5`:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --weight-decay 5e-5 \
  --wd-decay-epochs 300 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb_group dm-decaying-wd
```

---

## Reporting

Report `val_primary/surface_rel_l2_pct` at best epoch for both trials. Include W&B run IDs and note the epoch at which you achieved best validation. Also note at which epoch WD reached 0 and whether training dynamics visibly changed at that point (check the loss curves in W&B around epoch 300).

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shouko:

## Results

### Trial 1: WD=1e-4, linear decay over 300 epochs

| Metric | Baseline | Trial 1 |
|--------|----------|---------|
| `val_primary/surface_rel_l2_pct` | **3.833%** (ep511) | **4.602%** (ep230) |
| Best epoch | 511 | 230 |
| WD at best epoch | 1e-4 (constant) | 2.33e-5 |

- **W&B run**: `82v9rvo6`
- **Command**:
  ```
  cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
    --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
    --weight-decay 1e-4 --wd-decay-epochs 300 --enable-fourier \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
    --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 \
    --agent shouko --wandb-name "shouko/dm-wd1e4-decay300" --wandb-group dm-decaying-wd
  ```
- **Peak memory**: DDP on 8 GPUs, ~2.4 GB/GPU (same as baseline config)
- **Total epochs completed**: 300

### Trial 2: WD=5e-5, linear decay over 300 epochs

| Metric | Baseline | Trial 2 |
|--------|----------|---------|
| `val_primary/surface_rel_l2_pct` | **3.833%** (ep511) | **12.253%** (ep52) |
| Best epoch | 511 | 52 |

- **W&B run**: `s7gezo80`
- **Command**: Same as T1 but `--weight-decay 5e-5`, `CUDA_VISIBLE_DEVICES=1`, `--wandb-name "shouko/dm-wd5e5-decay300"`
- **Diverged catastrophically** at epoch 53. Grad norms jumped from ~1.0 to 1.6→6→28→258→Infinity by ep73. Killed at ep77 (val=54.2%).

### What happened

**Hypothesis not confirmed.** Neither trial beat the 3.833% baseline.

**Trial 1** showed excellent early convergence — reaching 4.6% by epoch 230 compared to the baseline needing 511 epochs for 3.833%. However, the linear decay to zero was destabilizing:

- **ep1–230**: Healthy training. Grad norms dropped from 9.9 (ep1) to 0.18 (ep230). Val improved monotonically.
- **ep230–235**: WD fell below ~2.2e-5. Grad norms doubled (0.18→0.36). Val stopped improving.
- **ep240–270**: Grad norms 0.4–1.1, val oscillating 4.7–6.9%.
- **ep280+**: Full divergence. Grad norm 134→Infinity. Val 25%→54%.
- **ep300 (WD=0)**: val=53.66%, grad_norm=Infinity.

The WD decay provided real regularization benefit early (smooth convergence, low grad norms) but decaying all the way to zero removed the weight norm control that AdamW relies on for stability. The critical WD threshold for this model appears to be ~2e-5.

**Trial 2** confirmed that WD=5e-5 is too weak for this model — it diverged almost immediately (ep53), never achieving a competitive val metric. The lower starting WD gave less regularization headroom, so the decay pushed it into instability much faster.

### Suggested follow-ups

1. **WD decay with a floor**: Decay from 1e-4 to a floor of 2.5e-5 or 3e-5 (instead of 0). This preserves the early convergence benefit while avoiding the instability region. The ep230 result (4.602%) suggests this schedule could reach ~3.8% or better given more epochs at a stable WD floor.
2. **Cosine WD decay**: Use cosine schedule `wd = initial_wd * 0.5 * (1 + cos(π * epoch / N))` which decays more slowly in the middle epochs, keeping WD higher for longer.
3. **Longer decay horizon**: Decay over 600+ epochs (past the expected convergence point) so WD is still meaningful when the model reaches its best performance.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3249 because ADVISOR: Closing — decaying WD destabilizes DM training. Trial 1 (WD=1e-4 → 0 over 300 eps) reached 4.602% at ep230 (decent convergence rate) but diverged when WD dropped below ~2e-5. Trial 2 (WD=5e-5 → 0) diverged at ep53. Useful finding: constant WD=1e-4 provides real regularization benefit during early training. This informs megumi's WD sweep (#3313, testing WD=1e-3 and WD=1e-2 as CONSTANT values on the champion config)..
